### PR TITLE
Deal with zero radius points in a morphology

### LINKED ISSUE #1526: Morphology causes NaN membrane values
Running the attached script with Arbor 02dd4326f0fdbb2e5c64f76a32c4f44320e9ee9d leads to NaN values for the membrane voltage.

Maybe related to `0` radius, cf. https://mcnest.slack.com/archives/C04NU4NMK/p1620833904155700.

[membrane_nan.py.gz](https://github.com/arbor-sim/arbor/files/6468161/membrane_nan.py.gz)

### PR DESCRIPTION
**Pull request by halfflat**
* Add test that is sensitive to small radius loss of precision in ixa computation.
* Add test that checks for consistent behaviour when there is an (isolated) zero radius point in the morphology.
* Special case p==1 q==1 case in ratelem so that it can interpolate ixa in the presence of non-finite interpolants.
* Change naming in pw_element/pw_elements: 'element' refers to the extent+value pairs comprising a 'pw_elements' object; 'value' refers to the value associated with an element; 'extent' refers to the closed interval which is the support of the element.
* Allow values (but not extents) in a pw_elements object to be mutable, using proxies for iterator access.
* Write piecewise zip (pw_zip) in terms of a lazy pw_zip_view. Rename zip functions.
* Document pw_elements more thoroughly.
* Simplify and document embed_pwlin.cpp routines, expressing everything in terms of simple interpolate/integrate operations.
* Represent integrated inverse cross-sectional area in the embedding by multiple piecewise-rational functions over a branch, each of which contribute separately to an ixa-based integration, so that precision loss associated with small (or zero) radii can be avoided.
* Comment embed_pwlin.cpp more thoroughly.

Fixes #1526

### CODE REVIEW

File: arbor/morph/embed_pwlin.cpp [hunk 1]
```diff
@@ -6,6 +6,8 @@
 #include <arbor/morph/morphology.hpp>
 #include <arbor/morph/primitives.hpp>
 
+#include "io/trace.hpp"
```
[the remark this answers is not in the document]
Comment (halfflat) [reply arbor/morph/embed_pwlin.cpp:9]: Not at all — an artefact from debugging. Will remove.

File: arbor/morph/embed_pwlin.cpp [hunk 2]
```diff
@@ -14,118 +16,72 @@
...
 template <unsigned p, unsigned q>
-using branch_pw_ratpoly = std::vector<pw_ratpoly<p, q>>;
+using pw_ratpoly = util::pw_elements<rat_element<p, q>>;
 
-// Special case handling required for degenerate branches of length zero:
-template <typename Elem>
-static bool is_degenerate(const util::pw_elements<Elem>& pw) {
-    return  pw.bounds().second==0;
-}
+// One branch can be convered by more than one pw_ratpoly. When the function
```
[the remark this answers is not in the document]
Comment (halfflat) [reply arbor/morph/embed_pwlin.cpp:39]: They must indeed; I can make this explicit.

File: arbor/util/ratelem.hpp [hunk 3]
```diff
@@ -73,29 +72,49 @@ double rat_eval(const std::array<double, 1+a+c>& g, const std::array<double, 2+a
...
+        // Special case when p==1 and q==1 so that we can have
+        // well-defined behaviour in the case where the interpolants
+        // are strictly monotonic, but might take non-finite values.
+        // (Used for ixa interpolation in embed_pwlin.cpp).
+
+        double y0 = data[0];
+        double y1 = data[1];
+        double y2 = data[2];
+
+        return y1+(2*x-1)/(x/(y2-y1)+(1-x)/(y1-y0));
```
[the remark this answers is not in the document]
Comment (halfflat) [reply arbor/util/ratelem.hpp:108]: It can be derived from the formula you get from the generic case, cancelling some terms. Or more directly, it's an order (1, 1) rational function f(x) (i.e. p(x)/q(x) where p and q are order 1 polynomials), and by inspection satisfies f(0) = y0, f(0.5) = y1, f(1) = y2, and so must be the desired interpolating function.

File: arbor/morph/embed_pwlin.cpp [hunk 4]
```diff
@@ -304,11 +355,52 @@ embed_pwlin::embed_pwlin(const arb::morphology& m) {
...
+                    ixa_0 = INFINITY;
+                    ixa_half = INFINITY;
+                    ixa_1 = INFINITY;
+                }
+                else if (r0==0) {
+                    ixa_0 = -INFINITY;
+                    ixa_half = -dx/(pi*r1*r1);
+                    ixa_1 = 0;
+                }
+
```
[the remark this answers is not in the document]
Comment (halfflat) [reply arbor/morph/embed_pwlin.cpp:375]: This is the case covered in lines 371-373; with r0==0, we pick the interpolating values so that we end up at zero at x==1, from a negative infinite start at x==0 (any finite offset added to this would also work, but this is the simplest).

File: arbor/util/piecewise.hpp [hunk 5]
```diff
@@ -1,15 +1,105 @@
 #pragma once
 
-// Create/manipulate 1-d piece-wise defined objects.
+// Create/manipulate 1-d piecewise defined objects.
 //
-// Using vectors everywhere here for ease; consider making
-// something more container/sequence-generic later.
+// A `pw_element<A>` describes a _value_ of type `A` and an _extent_ of
```
[the remark this answers is not in the document]
Comment (halfflat) [reply arbor/util/piecewise.hpp:5]: It's meant to capture a division of the domain into contiguous intervals, without an associated value. It's useful for zipping with another `pw_elements<A>` object, to constrain it to a subdomain for example. This is how it is used in `embed_pwlin.cpp` and `fvm_layout.cpp`.

File: arbor/util/piecewise.hpp [hunk 6]
```diff
@@ -1,15 +1,105 @@
...
+//
+// A `pw_element<A>` can be bound via a structured binding to an extent
+// and value, or just an extent if A is void.
+//
+// A `pw_elements<A>` object comprises a contiguous sequence Eᵢ of
+// `pw_element<A>` elements. If Eᵢ has extent [lᵢ, rᵢ] and Eᵢ₊₁ has extent
+// [lᵢ₊₁, rᵢ₊₁] then rᵢ must equal lᵢ₊₁.
+//
+// The `value_type` of `pw_elements<A>` is the type of the elements, i.e.
+// `pw_element<A>`, as `pw_elements<A>` presents as a read-only container.
```
[the remark this answers is not in the document]
Comment (halfflat) [reply arbor/util/piecewise.hpp:24]: Once upon a time it was read-only, and I forgot to update the comment :)

File: arbor/util/piecewise.hpp [hunk 7]
```diff
@@ -408,66 +646,80 @@ template <> struct pw_elements<void> {
...
-            return a_elem.element;
+// The piecewise map applies a transform to the values in a piecewise
+// object, preserving the extents. If the original piecewise object
+// is `pw_elements<void>`, then the function is called with zero arguments.
+//
+// If the mapping function returns void, return a `pw_elements<void>` object
+// with the same extents as the original piecewise object.
+
+template <typename X, typename Fn>
+auto pw_map(const pw_elements<X>& pw, Fn&& fn) {
```
[the remark this answers is not in the document]
Comment (halfflat) [reply arbor/util/piecewise.hpp:666]: It came up as possibly useful in the context of @thorstenhater's work on axial diffusion, so it has become happily relevant.

File: test/unit/test_morph_embedding.cpp [hunk 8]
```diff
@@ -286,3 +286,66 @@ TEST(embedding, area_0_length_segment) {
...
+    ASSERT_FALSE(std::isnan(computed_ixa));
+    EXPECT_TRUE(near_relative(expected_ixa, computed_ixa, reltol));
+
+    // Integrating over the zero radius point though should give us
+    // INFINITY.
+
+    double infinite_ixa = em.integrate_ixa(mcable{0, 0.25, 0.75});
+    ASSERT_FALSE(std::isnan(infinite_ixa));
+    EXPECT_TRUE(std::isinf(infinite_ixa));
+
```
[the remark this answers is not in the document]
Comment (halfflat) [reply test/unit/test_morph_embedding.cpp:337]: According to the comments in `embed_pwlin.cpp` on the integration functions, it should be infinite. But I'll add a test and see!

File: arbor/morph/embed_pwlin.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (noraabiakar) [reply arbor/morph/embed_pwlin.cpp:373]: I think I meant `(r1==0 && r0!=0)`, but I guess the default values of `ixa_0`, `ixa_half` and `ixa_1` hold for that case.

[the remark this answers is not in the document]
Comment (halfflat) [reply arbor/morph/embed_pwlin.cpp:373]: Yes — the code presumes IEEE semantics, which should give us `ixa_0==0`, `ixa_half==dx/(pi*r0*r0)`, and `ixa_1==INFINITY`.

**Review by noraabiakar — APPROVED** [review]
